sql: Self-referencing tables can now reference a non-primary index without manually adding an index.

BramGruneir opened the pull request:
Previously, this would fail:
```sql
CREATE TABLE test20042 (
  x STRING PRIMARY KEY
 ,y STRING UNIQUE
 ,z STRING REFERENCES test20042(y)
);
```

But this would pass:

```sql
CREATE TABLE test20042 (
  x STRING PRIMARY KEY
 ,y STRING UNIQUE
 ,z STRING REFERENCES test20042(y)
 ,INDEX(z)
);
```

Release Note: Fixed a bug that prevented creating self-referencing tables that referenced a non-primary index without manually including the index on the referencing column.

Fixes #20042.

BramGruneir committed:
sql: Self-referencing tables can now reference a non-primary index without manually adding an index.

Previously, this would fail:
```sql
CREATE TABLE test20042 (
  x STRING PRIMARY KEY
 ,y STRING UNIQUE
 ,z STRING REFERENCES test20042(y)
);
```

But this would pass:

```sql
CREATE TABLE test20042 (
  x STRING PRIMARY KEY
 ,y STRING UNIQUE
 ,z STRING REFERENCES test20042(y)
 ,INDEX(z)
);
```

Release Note: Fixed a bug that prevented creating self-referencing tables that referenced a non-primary index without manually including the index on the referencing column.

Fixes #20042.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20325)
<!-- Reviewable:end -->

vivekmenezes commented on pkg/sql/logictest/testdata/logic_test/fk:
```diff
@@ -874,15 +874,22 @@ pkref_b  CREATE TABLE pkref_b (
   x STRING PRIMARY KEY
- ,y STRING UNIQUE REFERENCES self_x2(x)
- ,z STRING REFERENCES self_x2(y)
- ,INDEX(z)
```
Why did you take out INDEX(z).

I'd actually not make the entire 20045 related change in this COMMIT

vivekmenezes commented on pkg/sql/logictest/testdata/logic_test/fk:
```diff
@@ -874,15 +874,22 @@ pkref_b  CREATE TABLE pkref_b (
-DELETE FROM self_x2 WHERE x = 'pk1';
+DELETE FROM test20045 WHERE x = 'pk1';
+
+# See https://github.com/cockroachdb/cockroach/issues/20042
```
no need for this since you already refer to 20042 in the table name

vivekmenezes commented on pkg/sql/logictest/testdata/logic_test/fk:
```diff
@@ -874,15 +874,22 @@ pkref_b  CREATE TABLE pkref_b (
+  x STRING PRIMARY KEY
+ ,y STRING UNIQUE
+ ,z STRING REFERENCES test20042(y)
+);
```
Wanna add an INSERT and DELETE after this to make sure that it works for this schema

vivekmenezes commented on pkg/sql/create_table.go:
```diff
@@ -403,15 +403,22 @@ func resolveFK(
+	// the index's index into target's list of indexes. If it is a primary index,
+	// targetIdxIndex is set to -1. Also store the targetIndex's ID so we
+	// don't have to do the lookup twice.
+	targetIdxIndex := -1
```
I think you can get rid of this and use only targetIdxID in this change

vivekmenezes reviewed (commented):
Thanks for fixing this!

BramGruneir committed:
sql/logictest: small update to test for #20045

Since #20042 has been fixes, the extra index is no longer needed. Also added some extra rows and deletes for good measure.

BramGruneir commented:
I took out the `INDEX(z)` as it was only required for table creation due to this bug.  I've split this into two commits, the second one dealing with 20045 exclusively.

---

Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/create_table.go, line 411 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20325#-L-7PUCrEZbD1uVTo8EH-r1-411:-L-Dld6jJeYyOOejNuIP:b-zi9i6d) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b5345017845b03d24359bb5f44836b554fe271b/pkg/sql/create_table.go#L411)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

I think you can get rid of this and use only targetIdxID in this change
</blockquote></details>

How?  We have an slice of indices, but I don't think that the location in the slice matches the ID every time.

---

*[pkg/sql/logictest/testdata/logic_test/fk, line 889 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20325#-L-7PUCrEZbD1uVTo8EI-r1-889:-L-DlyEBoxT4On3H_OBJ:bd0bfwx) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b5345017845b03d24359bb5f44836b554fe271b/pkg/sql/logictest/testdata/logic_test/fk#L889)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

no need for this since you already refer to 20042 in the table name
</blockquote></details>

Sure, I'll remove the comment, just seemed harmless to keep in there and matches how other issues are referenced in the logic tests.

---

*[pkg/sql/logictest/testdata/logic_test/fk, line 895 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20325#-L-7PUCrEZbD1uVTo8EI-r1-895:-L-DmwpMEl-yudrDofl9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b5345017845b03d24359bb5f44836b554fe271b/pkg/sql/logictest/testdata/logic_test/fk#L895)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Wanna add an INSERT and DELETE after this to make sure that it works for this schema
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20325#-:-L-DnB4BxKQbYbKuw_o3:b1xjotj)*
<!-- Sent from Reviewable.io -->
